File: test/XRoadLib.Tests/Protocols/Headers/XRoadIdentifierTest.fs
namespace XRoadLib.Tests

open FsUnit
open NUnit.Framework
open System
open XRoadLib.Protocols.Headers

[<TestFixture>]
module XRoadClientIdentifierTest =
    let [<Test>] ``identifier without content`` () =
        let id = XRoadClientIdentifier()
        id.ToString() |> should equal ""

    let [<Test>] ``only X-Road instance`` () =
        let id = XRoadClientIdentifier(XRoadInstance = "ee-dev")
        id.ToString() |> should equal "ee-dev//"

    let [<Test>] ``only member class`` () =
        let id = XRoadClientIdentifier(MemberClass = "GOV")
        id.ToString() |> should equal "/GOV/"

    let [<Test>] ``old header identifier (member code only)`` () =
        let id = XRoadClientIdentifier(MemberCode = "11111111")
        id.ToString() |> should equal "11111111"

    let [<Test>] ``only subsystem`` () =
        let id = XRoadClientIdentifier(SubsystemCode = "test")
        id.ToString() |> should equal "///test"

    let [<Test>] ``has parts of new X-Road identifier and subsystem code`` () =
        let id = XRoadClientIdentifier(XRoadInstance = "ee-dev", SubsystemCode = "test")
        id.ToString() |> should equal "ee-dev///test"

    let [<Test>] ``full identifier`` () =
        let id = XRoadClientIdentifier(XRoadInstance = "ee-dev", MemberClass = "GOV", MemberCode = "11111111", SubsystemCode = "test")
        id.ToString() |> should equal "ee-dev/GOV/11111111/test"

[<TestFixture>]
module XRoadCentralServiceIdentifierTest =
    let [<Test>] ``empty identifier`` () =
        XRoadCentralServiceIdentifier().ToString() |> should equal "/"

    let [<Test>] ``only X-Road instance`` () =
        XRoadCentralServiceIdentifier(XRoadInstance = "ee-dev").ToString() |> should equal "ee-dev/"

    let [<Test>] ``only service code`` () =
        XRoadCentralServiceIdentifier(ServiceCode = "test").ToString() |> should equal "/test"

    let [<Test>] ``full identifier`` () =
        XRoadCentralServiceIdentifier(XRoadInstance = "ee-dev", ServiceCode = "test").ToString() |> should equal "ee-dev/test"

[<TestFixture>]
module XRoadServiceNameTest =
    let [<Literal>] producerName = "producerName"
    let [<Literal>] serviceName = "serviceName"
    let [<Literal>] version = 3u

    let fullName = sprintf "%s.%s.v%d" producerName serviceName version

    [<Test>]
    let ``can parse full service name`` () =
        let xsn = XRoadServiceIdentifier.FromString(fullName)
        xsn.SubsystemCode |> should be Null
        xsn.ServiceCode |> should equal serviceName
        xsn.Version |> should equal version

    [<Test>]
    let ``initialize with null value`` () =
        let xsn = XRoadServiceIdentifier.FromString(null)
        xsn |> should not' (be Null)
        xsn.SubsystemCode |> should be Null
        xsn.ServiceCode |> should be Null
        xsn.Version |> should be Null

    [<Test>]
    let ``initialize with empty string`` () =
        let xsn = XRoadServiceIdentifier.FromString("")
        xsn.SubsystemCode |> should be Null
        xsn.ServiceCode |> should equal ""
        xsn.Version |> should be Null

    [<Test>]
    let ``initialize without producer in full name`` () =
        let withoutProducer = "serviceName.v1"
        let xsn = XRoadServiceIdentifier.FromString(withoutProducer)
        xsn.SubsystemCode |> should be Null
        xsn.ServiceCode |> should equal "serviceName"
        xsn.Version |> should equal (Nullable(1u))
        xsn.ToFullName() |> should equal "serviceName.v1"
        xsn.ToString() |> should equal "/serviceName/v1"

    [<Test>]
    let ``initialize without producer and version in full name`` () =
        let withoutProducerAndVersion = "serviceName"
        let xsn = XRoadServiceIdentifier.FromString(withoutProducerAndVersion)
        xsn.SubsystemCode |> should be Null
        xsn.ServiceCode |> should equal "serviceName"
        xsn.Version |> should be Null
        xsn.ToFullName() |> should equal "serviceName"
        xsn.ToString() |> should equal "/serviceName"

    [<Test>]
    let ``initialize without version in full name`` () =
        let withoutVersion = "producer.listMethods"
        let xsn = XRoadServiceIdentifier.FromString(withoutVersion)
        xsn.SubsystemCode |> should be Null
        xsn.ServiceCode |> should equal "listMethods"
        xsn.Version |> should be Null
        xsn.ToFullName() |> should equal "listMethods"
        xsn.ToString() |> should equal "/listMethods"

    [<Test>]
    let ``initialize without serviceName in full name`` () =
        let withoutServiceName = "producer..v2"
        let xsn = XRoadServiceIdentifier.FromString(withoutServiceName)
        xsn.SubsystemCode |> should be Null
        xsn.ServiceCode |> should equal ""
        xsn.Version |> should equal (Nullable(2u))
        xsn.ToFullName() |> should equal ".v2"
        xsn.ToString() |> should equal "//v2"

    [<Test>]
    let ``initialize with version 0`` () =
        let fullName = sprintf "%s.%s.v%d" producerName serviceName 0u
        let xsn = XRoadServiceIdentifier.FromString(fullName)
        xsn.SubsystemCode |> should be Null
        xsn.ServiceCode |> should equal "serviceName"
        xsn.Version |> should equal (Nullable(0u))
        xsn.ToFullName() |> should equal "serviceName.v0"
        xsn.ToString() |> should equal "/serviceName/v0"

    [<Test>]
    let ``convert to string with all parts assigned`` () =
        let xsn = XRoadServiceIdentifier.FromString(fullName)
        xsn.ToFullName() |> should equal "serviceName.v3"
        xsn.ToString() |> should equal "/serviceName/v3"

[<TestFixture>]
module XRoadServiceIdentifierTest =
    let [<Test>] ``empty identifier`` () =
        XRoadServiceIdentifier().ToString() |> should equal "/"

    let [<Test>] ``only version`` () =
        XRoadServiceIdentifier(ServiceVersion = "v1").ToString() |> should equal "//v1"

    let [<Test>] ``only service code`` () =
        XRoadServiceIdentifier(ServiceCode = "test").ToString() |> should equal "/test"

    let [<Test>] ``service code and version`` () =
        XRoadServiceIdentifier(ServiceCode = "test", ServiceVersion = "v1").ToString() |> should equal "/test/v1"

    let [<Test>] ``subsystem code only`` () =
        XRoadServiceIdentifier(SubsystemCode = "generic-producer").ToString() |> should equal "generic-producer/"

    let [<Test>] ``subsystem code with service and version`` () =
        XRoadServiceIdentifier(SubsystemCode = "generic", ServiceCode = "test", ServiceVersion = "v1").ToString() |> should equal "generic/test/v1"

    let [<Test>] ``subsystem code with version`` () =
        XRoadServiceIdentifier(SubsystemCode = "generic", ServiceVersion = "v1").ToString() |> should equal "generic//v1"

    let [<Test>] ``subsystem code with service`` () =
        XRoadServiceIdentifier(SubsystemCode = "generic", ServiceCode = "test").ToString() |> should equal "generic/test"

    let [<Test>] ``X-Road instance only`` () =
        XRoadServiceIdentifier(XRoadInstance = "ee-dev").ToString() |> should equal "ee-dev////"

    let [<Test>] ``member class only`` () =
        XRoadServiceIdentifier(MemberClass = "GOV").ToString() |> should equal "/GOV///"

    let [<Test>] ``member code only`` () =
        XRoadServiceIdentifier(MemberCode = "11111111").ToString() |> should equal "//11111111//"

    let [<Test>] ``full identifier`` () =
        XRoadServiceIdentifier(XRoadInstance = "ee-dev", MemberClass = "GOV", MemberCode = "11111111", SubsystemCode = "generic", ServiceCode = "test", ServiceVersion = "v1").ToString() |> should equal "ee-dev/GOV/11111111/generic/test/v1"

    let [<Test>] ``full identifier without subsystem code`` () =
        XRoadServiceIdentifier(XRoadInstance = "ee-dev", MemberClass = "GOV", MemberCode = "11111111", ServiceCode = "test", ServiceVersion = "v1").ToString() |> should equal "ee-dev/GOV/11111111//test/v1"

File: test/XRoadLib.Tests/Protocols/Headers/XRoadIdentifierTest.cs
using XRoadLib.Headers;

namespace XRoadLib.Tests.Protocols.Headers;

public class XRoadIdentifierTest
{
    private const string ProducerName = "producerName";
    private const string ServiceName = "serviceName";
    private const uint Version = 3u;

    private readonly string _fullName = $"{ProducerName}.{ServiceName}.v{Version}";

    [Theory]
    [InlineData(null, null, null, null, "")]
    [InlineData("ee-dev", null, null, null, "ee-dev//")]
    [InlineData(null, "GOV", null, null, "/GOV/")]
    [InlineData(null, null, "11111111", null, "11111111")]
    [InlineData(null, null, null, "test", "///test")]
    [InlineData("ee-dev", null, null, "test", "ee-dev///test")]
    [InlineData("ee-dev", "GOV", "11111111", "test", "ee-dev/GOV/11111111/test")]
    public void XRoadClientIdentifierTest(string instance, string memberClass, string memberCode, string subsystemCode, string expectedValue)
    {
        var id = new XRoadClientIdentifier
        {
            XRoadInstance = instance,
            MemberClass = memberClass,
            MemberCode = memberCode,
            SubsystemCode = subsystemCode
        };
        Assert.Equal(expectedValue, id.ToString());
    }

    [Theory]
    [InlineData(null, null, "/")]
    [InlineData("ee-dev", null, "ee-dev/")]
    [InlineData(null, "test", "/test")]
    [InlineData("ee-dev", "test", "ee-dev/test")]
    public void XRoadCentralServiceIdentifierTest(string instance, string serviceCode, string expectedValue)
    {
        var id = new XRoadCentralServiceIdentifier
        {
            XRoadInstance = instance,
            ServiceCode = serviceCode
        };
        Assert.Equal(expectedValue, id.ToString());
    }

    [Theory]
    [InlineData(null, null, null, null, null, null, "/")]
    [InlineData(null, null, null, null, null, "v1", "//v1")]
    [InlineData(null, null, null, null, "test", null, "/test")]
    [InlineData(null, null, null, null, "test", "v1", "/test/v1")]
    [InlineData(null, null, null, "generic-producer", null, null, "generic-producer/")]
    [InlineData(null, null, null, "generic", "test", "v1", "generic/test/v1")]
    [InlineData(null, null, null, "generic", null, "v1", "generic//v1")]
    [InlineData(null, null, null, "generic", "test", null, "generic/test")]
    [InlineData("ee-dev", null, null, null, null, null, "ee-dev////")]
    [InlineData(null, "GOV", null, null, null, null, "/GOV///")]
    [InlineData(null, null, "11111111", null, null, null, "//11111111//")]
    [InlineData("ee-dev", "GOV", "11111111", "generic", "test", "v1", "ee-dev/GOV/11111111/generic/test/v1")]
    [InlineData("ee-dev", "GOV", "11111111", null, "test", "v1", "ee-dev/GOV/11111111//test/v1")]
    public void XRoadServiceIdentifierTest(string instance, string memberClass, string memberCode, string subsystemCode, string serviceCode, string serviceVersion, string expectedValue)
    {
        var id = new XRoadServiceIdentifier
        {
            XRoadInstance = instance,
            MemberClass = memberClass,
            MemberCode = memberCode,
            SubsystemCode = subsystemCode,
            ServiceCode = serviceCode,
            ServiceVersion = serviceVersion
        };
        Assert.Equal(expectedValue, id.ToString());
    }

    [Fact]
    public void CanParseFullServiceName()
    {
        var xsn = XRoadServiceIdentifier.FromString(_fullName);
        Assert.Null(xsn.SubsystemCode);
        Assert.Equal(ServiceName, xsn.ServiceCode);
        Assert.Equal(Version, xsn.Version);
    }

    [Fact]
    public void CanParseNullValue()
    {
        var xsn = XRoadServiceIdentifier.FromString(null);
        Assert.NotNull(xsn);
        Assert.Null(xsn.SubsystemCode);
        Assert.Null(xsn.ServiceCode);
        Assert.Null(xsn.Version);
    }

    [Fact]
    public void CanParseEmptyString()
    {
        var xsn = XRoadServiceIdentifier.FromString("");
        Assert.Null(xsn.SubsystemCode);
        Assert.Equal("", xsn.ServiceCode);
        Assert.Null(xsn.Version);
    }

    [Fact]
    public void InitializeWithoutProducerInFullName()
    {
        const string withoutProducer = "serviceName.v1";
        var xsn = XRoadServiceIdentifier.FromString(withoutProducer);
        Assert.Null(xsn.SubsystemCode);
        Assert.Equal("serviceName", xsn.ServiceCode);
        Assert.Equal(1u, xsn.Version);
        Assert.Equal("serviceName.v1", xsn.ToFullName());
        Assert.Equal("/serviceName/v1", xsn.ToString());
    }

    [Fact]
    public void InitializeWithoutProducerAndVersionInFullName()
    {
        const string withoutProducerAndVersion = "serviceName";
        var xsn = XRoadServiceIdentifier.FromString(withoutProducerAndVersion);
        Assert.Null(xsn.SubsystemCode);
        Assert.Equal("serviceName", xsn.ServiceCode);
        Assert.Null(xsn.Version);
        Assert.Equal("serviceName", xsn.ToFullName());
        Assert.Equal("/serviceName", xsn.ToString());
    }

    [Fact]
    public void InitializeWithoutVersionInFullName()
    {
        const string withoutVersion = "producer.listMethods";
        var xsn = XRoadServiceIdentifier.FromString(withoutVersion);
        Assert.Null(xsn.SubsystemCode);
        Assert.Equal("listMethods", xsn.ServiceCode);
        Assert.Null(xsn.Version);
        Assert.Equal("listMethods", xsn.ToFullName());
        Assert.Equal("/listMethods", xsn.ToString());
    }

    [Fact]
    public void InitializeWithoutServiceNameInFullName()
    {
        const string withoutServiceName = "producer..v2";
        var xsn = XRoadServiceIdentifier.FromString(withoutServiceName);
        Assert.Null(xsn.SubsystemCode);
        Assert.Equal("", xsn.ServiceCode);
        Assert.Equal((uint?)2, xsn.Version);
        Assert.Equal(".v2", xsn.ToFullName());
        Assert.Equal("//v2", xsn.ToString());
    }

    [Fact]
    public void InitializeWithVersion0()
    {
        var fullName = $"{ProducerName}.{ServiceName}.v0";
        var xsn = XRoadServiceIdentifier.FromString(fullName);
        Assert.Null(xsn.SubsystemCode);
        Assert.Equal("serviceName", xsn.ServiceCode);
        Assert.Equal((uint?)0, xsn.Version);
        Assert.Equal("serviceName.v0", xsn.ToFullName());
        Assert.Equal("/serviceName/v0", xsn.ToString());
    }

    [Fact]
    public void ConvertToStringWithAllPartsAssigned()
    {
        var xsn = XRoadServiceIdentifier.FromString(_fullName);
        Assert.Equal("serviceName.v3", xsn.ToFullName());
        Assert.Equal("/serviceName/v3", xsn.ToString());
    }
}